File: examples/function_repair/repair_function_trace_test.py
from pathlib import Path
from tempfile import TemporaryDirectory

from repair_function_package import repair_function_agent


def test_repairs_one_named_python_function_through_tools():
    with TemporaryDirectory() as tmp:
        source_path = Path(tmp) / "billing.py"
        source_path.write_text("def calculate_total(subtotal: int, discount: int) -> int:\n    return subtotal\n", encoding="utf8")
        trace: list[str] = []

        class Tools:
            def read_text(self, path: str) -> str:
                trace.append(f"read:{path}")
                return Path(path).read_text(encoding="utf8")

            def search_text(self, query: str) -> list[str]:
                trace.append(f"search:{query}")
                return [str(source_path)]

            def write_text(self, path: str, content: str) -> dict[str, object]:
                trace.append(f"edit:{path}")
                Path(path).write_text(content, encoding="utf8")
                return {"ok": True, "output": "wrote billing.py"}

            def run_test(self, command: str) -> dict[str, object]:
                trace.append(f"test:{command}")
                return {"ok": True, "output": "billing test passed"}

        agent = repair_function_agent(tools=Tools())
        result = agent.run({
            "path": str(source_path),
            "function_name": "calculate_total",
            "desired_behavior": "Return subtotal minus discount.",
            "test_command": "pytest tests/test_billing.py",
        })

        assert result.ok
        assert result.output["function_name"] == "calculate_total"
        assert result.output["test_passed"] is True
        assert result.output["edited"] is True
        assert "subtotal - discount" in source_path.read_text(encoding="utf8")
        assert trace == [
            f"read:{source_path}",
            "search:calculate_total",
            f"edit:{source_path}",
            "test:pytest tests/test_billing.py",
        ]


if __name__ == "__main__":
    test_repairs_one_named_python_function_through_tools()

File: examples/function_repair/repair_function_trace.test.ts
import { mkdtemp, readFile, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { repairFunctionAgent } from "./repair_function_package.js";

describe("repair_function golden workflow trace", () => {
  it("repairs one named TypeScript function through read/search/edit/test tools", async () => {
    const dir = await mkdtemp(join(tmpdir(), "forma-repair-ts-"));
    const sourcePath = join(dir, "billing.ts");
    const trace: string[] = [];
    await writeFile(sourcePath, "export function calculateTotal(subtotal: number, discount: number): number {\n  return subtotal;\n}\n", "utf8");

    const agent = repairFunctionAgent({
      readText: async (path) => {
        trace.push(`read:${path}`);
        return readFile(path, "utf8");
      },
      searchText: async (query) => {
        trace.push(`search:${query}`);
        return [sourcePath];
      },
      writeText: async (path, content) => {
        trace.push(`edit:${path}`);
        await writeFile(path, content, "utf8");
        return { ok: true, output: "wrote billing.ts" };
      },
      runTest: async (command) => {
        trace.push(`test:${command}`);
        return { ok: true, output: "billing test passed" };
      },
    });

    const result = await agent.run({
      path: sourcePath,
      function_name: "calculateTotal",
      desired_behavior: "Return subtotal minus discount.",
      test_command: "pnpm test -- billing",
    });

    expect(result.ok).toBe(true);
    expect(result.output).toMatchObject({
      function_name: "calculateTotal",
      test_passed: true,
      edited: true,
    });
    expect(await readFile(sourcePath, "utf8")).toContain("subtotal - discount");
    expect(trace).toEqual([
      `read:${sourcePath}`,
      "search:calculateTotal",
      `edit:${sourcePath}`,
      "test:pnpm test -- billing",
    ]);
  });
});
